router/middleware/perm/perm.go

```diff
@@ -87,21 +88,39 @@ func MustSecretAdmin() gin.HandlerFunc {
 				return
 			}
 		case constants.SecretShared:
-			logrus.Debugf("Verifying user %s has 'admin' permissions for team %s/%s", u.GetName(), o, n)
+			if n == "*" && m == "GET" {
+				logrus.Debugf("Verifying user %s has 'admin' permissions for org %s", u.GetName(), o)
 
-			perm, err := source.FromContext(c).TeamAccess(u, o, n)
-			if err != nil {
-				logrus.Errorf("unable to get user %s access level for team %s/%s: %v", u.GetName(), o, n, err)
-			}
+				perm, err := source.FromContext(c).OrgAccess(u, o)
+				if err != nil {
+					logrus.Errorf("unable to get user %s access level for org %s: %v", u.GetName(), o, err)
+				}
 
-			if !strings.EqualFold(perm, "admin") {
-				// nolint: lll // ignore long line length due to error message
-				retErr := fmt.Errorf("user %s does not have 'admin' permissions for the team %s/%s", u.GetName(), o, n)
+				if !strings.EqualFold(perm, "admin") {
+					// nolint: lll // ignore long line length due to error message
+					retErr := fmt.Errorf("user %s does not have 'admin' permissions for the org %s", u.GetName(), o)
 
-				util.HandleError(c, http.StatusUnauthorized, retErr)
+					util.HandleError(c, http.StatusUnauthorized, retErr)
 
-				return
+					return
+				}
```

Comment:
I might be misunderstanding what the above code is doing.

But if I'm reading this correctly, I think this would break the perms contract Vela has for `shared` secrets?

Looking at this change, it appears it would allow an org `admin` to get a list of `shared` secrets.

The problem with that is Vela always required a user to be a team `admin` to access `shared` secrets.

The idea or thought process is `shared` secrets are meant to be scoped to a specific team in the org.

Reply:
Wouldn't an `org` admin inherit admin for the teams? I think in GitHub `org` admins have full ability to manage teams under their `org`.

On the other hand if you are not an `org` admin you won't see any `shared` secrets in the `org` page. Which really isn't any different from today where you have to go directly to the shared secrets URL.

I'm not sure how else we would approach getting a list of shared secrets without abusing the GitHub api. One idea would be to collect some metadata about the user on login and store it in vela; it would be stale as soon as they made a change in GitHub. Maybe a refresh metadata route, that could be triggered on demand would alleviate that.

Metadata could include `orgs` the user belongs to, `teams` they are connected to along with `permissions`.

Reply:
> Wouldn't an org admin inherit admin for the teams? I think in GitHub org admins have full ability to manage teams under their org.

We could utilize the [/orgs/{org}/teams](https://docs.github.com/en/rest/reference/teams) to determine which team(s) the given user has access.

> I'm not sure how else we would approach getting a list of shared secrets without abusing the GitHub api.

Do we think that the average organization has many hundreds or thousands of teams? Looks like the previously mentioned endpoint can return 100 teams within the first request.

Reply:
> Wouldn't an `org` admin inherit admin for the teams?

No, I don't believe this is the case.

IIRC, we've designed this as intentional functionality in the system for specific use-cases.

> I think in GitHub `org` admins have full ability to manage teams under their `org`.

Yes I believe this is correct 👍 

An `org` admin can manage any `team` under that `org`.

---

Now I understand the above two statements might seem conflicting so I'm going to elaborate.

You can be an `org` admin, but still not have access to directly manage `shared` secrets for a `team` in that `org`.

Whenever a user attempts to access `shared` secrets, Vela specifically checks if they're an `admin` of the `team`:

https://github.com/go-vela/server/blob/e4801f81021eba1ab3d4909a57979c9a4b32365f/router/middleware/perm/perm.go#L89-L104

If they don't have access to the `team`, but they're an `org` admin, they'll face the following error:

```
user <user> does not have 'admin' permissions for the team <org>/<team>
```

However, like you pointed out above, an `org` admin can manage any `team` in that `org`.

This means that a user who is an `org` admin but isn't on a `team` in that `org` could go add themselves to the `team`.

Then, this would grant them the required access level to manage `shared` secrets for that `team` in Vela.

Reply:
@JayCeeJr and I hopped into a quick Slack call to discuss some of this 👍 

Based off what @JordanSussman mentioned above, we think we've found a good approach to enable this functionality.

This approach may use either the endpoint Jordan mentioned or the [`/user/teams`](https://docs.github.com/en/rest/reference/teams#list-teams-for-the-authenticated-user) endpoint.

Thinking this will help check what `teams` in an `org` the user has access to.

Then, we only have to list the `shared` secrets in those specific `teams`.

Reply:
> This approach may use either the endpoint Jordan mentioned or the /user/teams endpoint.

I'm thinking the `/user/teams` endpoint will be the most performant option.